internal/appsec/limiter.go

```diff
@@ -0,0 +1,114 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+//go:build appsec
+// +build appsec
+
+package appsec
+
+import (
+	"sync/atomic"
+	"time"
+)
+
+type Limiter interface {
+	Allow() bool
+}
+
+//The token ticker is a thread-safe and lock-free rate limiter based on a token bucket.
+//The idea is to have a goroutine that will update  the bucket with fresh tokens at regular intervals using a time.Ticker.
+//The advantage of using a goroutine here is  that the implementation becomes easily thread-safe using a few
+//atomic operations with little overhead overall. TokenTicker.Start() *must* be called before the first call to
+//TokenTicker.Allow() and TokenTicker.Stop() *must* be called once done using.
+type TokenTicker struct {
+	tokens    int64
+	maxTokens int64
+	ticker    *time.Ticker
+	stopChan  chan struct{}
+}
+
+func NewTokenTicker(tokens, maxTokens int64) *TokenTicker {
+	return &TokenTicker{
+		tokens:    tokens,
+		maxTokens: maxTokens,
+		stopChan:  make(chan struct{}),
+	}
+}
+
+//Select loop to update the token amount in the bucket. Used in a goroutine by the rate limiter.
+func (t *TokenTicker) updateBucket(startTime time.Time) {
+	ticksPerToken := (time.Second.Nanoseconds() / t.maxTokens)
+	ticks := int64(0)
+	prevStamp := startTime
+
+	for {
```

Comment:
Let's add a few descriptive comments in this for loop to help anyone better understand what it does and what the overall logic is please.